File: julia/ip.jl
using Memento
using JSON
using HTTP
using URIs

# To use included Project.toml to set up environment before running script, uncomment thsese lines
# using Pkg
# Pkg.activate(@__DIR__)
# Pkg.instantiate()

# Set up logging
logger = Memento.config!("info"; fmt="[{date}]: {msg}")
formatter = DefaultFormatter("[{date}]: {msg}"; date_fmt_string="U dd, yyyy HH:MM:SS")

handler = DefaultHandler("gddnsjl.log")
handler.fmt = formatter
push!(logger, handler)

# Function to parse config file
function config()
    path = joinpath(dirname(@__DIR__), "config.json")
    config = JSON.parsefile(path)
    return config
end

function main()
    ip = HTTP.get("https://domains.google.com/checkip").body |> x -> String(x)
    c = config()
    url = URI(parse(URI, "https://$(c["username"]):$(c["password"])@domains.google.com/nic/update"), query = Dict{String, Any}("hostname" => c["hostname"], "myip" => ip))
    rp = HTTP.post(url)
    info(logger, "hostname: $(c["hostname"]) IP: $(ip) response: $(String(rp.body))")
end

main()

File: python/ip.py
import logging
import json
import requests
import os
from pathlib import Path

"""
Set up logging
"""
logging.basicConfig(
    format="%(asctime)s -  %(message)s", datefmt="%d-%b-%y %H:%M:%S",
)
logger = logging.getLogger("googleddns")
fhandler = logging.FileHandler("gddnspy.log")
fformat = logging.Formatter("%(asctime)s -  %(message)s", datefmt="%d-%b-%y %H:%M:%S")
fhandler.setFormatter(fformat)
fhandler.setLevel("INFO")
logger.setLevel("INFO")
logger.addHandler(fhandler)


def config():
    path = Path(os.path.realpath(__file__))
    path = path.parent.parent.joinpath("config.json")
    with open(path, "rb") as f:
        config = json.loads(f.read())
    return config


def main():
    r = requests.get("https://domains.google.com/checkip")
    ip = r.content.decode(r.encoding)
    c = config()
    url = f"https://{c['username']}:{c['password']}@domains.google.com/nic/update"
    rp = requests.post(url, {"hostname": c["hostname"], "myip": ip})
    rp.raise_for_status()
    logger.info(
        "hostname: %s IP: %s response: %s"
        % (c["hostname"], ip, rp.content.decode(rp.encoding))
    )


if __name__ == "__main__":
    main()
